lib/cog/template.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Cog.Template do
+  alias Cog.Queries
+  alias Cog.Repo
+  alias Cog.TemplateCache
+
+  @fallback_adapter "any"
+
+  def render(adapter, bundle_id, template, context) do
+    with {:ok, template_fun} <- fetch_compiled_fun(adapter, bundle_id, template, context) do
+      template_fun.(context)
+    end
+  end
+
+  def fetch_compiled_fun(adapter, bundle_id, template, context) do
+    with :error              <- TemplateCache.lookup(adapter, bundle_id, template),
+         {:ok, source}       <- fetch_source(adapter, bundle_id, template, context),
+         {:ok, template_fun} <- compile(source),
+         :ok                 <- TemplateCache.insert(adapter, bundle_id, template, template_fun),
+         do: {:ok, template_fun}
+  end
+
+  def fetch_source(adapter, bundle_id, nil, context) do
+    fetch_source(adapter, bundle_id, default_template(context), context)
+  end
+
+  # Always use the raw template when responding to the test adapter.
+  # Used in integration tests.
+  def fetch_source("test", _bundle_id, _template, context) do
```

Comment:
Oh, yep. The only difference is that we're using `raw` here instead of `json`, which is easier to parse. It's not perfect, but I'm ok merging this and removing it once we have a better way to test commands in isolation.